tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoResultParser.java

```diff
@@ -0,0 +1,119 @@
+            .build());
+
+    //test result: FAILED. 25 passed; 1 failed; 0 ignored; 0 measured
+    private static final Pattern RESULT
+            = Pattern.compile("test result: .*\\. (?<passes>\\d*) passed; (?<fails>\\d*) failed; \\d* ignored; \\d* measured");
+    //test test::dim4::test_4d_1_80_perioditic ... ok
+    private static final Pattern TEST = Pattern.compile("test (?<name>.*) \\.\\.\\. .*");
+    //thread 'test::dim2::test_2d_1_80_normal' panicked at 'assertion failed: false', src\test\dim2.rs:12
+    private static final Pattern FAILURES
+            = Pattern.compile(".*thread '(?<name>.*)' panicked at '(?<description>.*)', .*");
+
+    /**
+     * Parses given process results outputs to a run result.
+     */
+    public RunResult parse(ProcessResult processResult) {
+        String output = processResult.output;
+        String[] lines = output.split("\\r?\\n");
+
+        Matcher matcher = RESULT.matcher(output);
+        while (matcher.find()) {
+            int fails = Integer.parseInt(matcher.group("fails"));
+            int passes = Integer.parseInt(matcher.group("passes"));
+            if (fails + passes > 0) {
+                Optional<Map<String, String>> failures
+                        = findFailures(processResult.output, fails);
+                if (failures.isPresent()) {
+                    Status status = fails == 0 ? Status.PASSED : Status.TESTS_FAILED;
+                    List<String> tests = parseResults(lines);
+                    return new RunResult(
+                            status,
+                            buildTestResults(tests, failures.get()),
+                            new ImmutableMap.Builder()
+                            .put(SpecialLogs.STDOUT,
+                                    processResult.output.getBytes(StandardCharsets.UTF_8))
+                            .put(SpecialLogs.STDERR,
+                                    processResult.errorOutput.getBytes(StandardCharsets.UTF_8))
+                            .build());
+                }
+                break;
+            }
+        }
+        return PARSING_FAILED;
+    }
+
+    private List<String> parseResults(String[] lines) {
+        List<String> result = new ArrayList<>();
+        for (String line : lines) {
+            Matcher matcher = TEST.matcher(line);
+            if (matcher.matches()) {
+                result.add(matcher.group("name"));
+            }
+        }
+        return result;
+    }
+
+    private Optional<Map<String, String>> findFailures(String errorOutput, int fails) {
+        Map<String, String> result = new HashMap<>();
+        String[] lines = errorOutput.split("\\r?\\n");
+        for (String line : lines) {
+            Matcher matcher = FAILURES.matcher(line);
+            if (matcher.matches()) {
+                result.put(matcher.group("name"), matcher.group("description"));
+            }
+        }
+        return Optional.of(result);
```

Comment:
Will never return `Optional.absent()`, since `result` is always non-null. Unwrap from `Optional`?